### PR TITLE
WI #1673 Fix continuation group scanning in incremental mode

### LINKED ISSUE #1673: Continuation lines
# Describe the bug
Generally speaking, continuation lines seem not to work well with incremental parsing and completion :
- real-time diagnostics are reported on continuation line but disappear after full reparsing (Ctrl-S)
- completion is broken on a continuation line (either no results or out of context results are returned)

# To Reproduce
From this code:
```cobol
       IDENTIFICATION DIVISION.
       PROGRAM-ID. MyPGM.
       DATA DIVISION.
       WORKING-STORAGE SECTION.
       01 Var1 pic X(80) value 'a'.
       END PROGRAM MyPGM.
```

Change it first to:
```cobol
       IDENTIFICATION DIVISION.
       PROGRAM-ID. MyPGM.
       DATA DIVISION.
       WORKING-STORAGE SECTION.
       01 Var1 pic X(80) value 'a
       END PROGRAM MyPGM.
```
A first error appear on the line of `Var1`: `Syntax error : Literal is not correctly delimited.`

Then change the code to:
```cobol
       IDENTIFICATION DIVISION.
       PROGRAM-ID. MyPGM.
       DATA DIVISION.
       WORKING-STORAGE SECTION.
       01 Var1 pic X(80) value 'a
      -     'aaa'. 
       END PROGRAM MyPGM.
```
A second error appear on the new line:
`Syntax error : extraneous input ''aaa'' expecting {separator, statement starting keyword, keyword}`

Now save the document which trigger a full refresh of the document.
The 2 errors disappear.

# Expected behavior
The 2 errors must disappear as soon as the second line is inserted.

# Others tests
## Test 1
Test 1 is the example above.

## Test 2
From this code:
```cobol
       IDENTIFICATION DIVISION.
       PROGRAM-ID. MyPGM.
       DATA DIVISION.
       WORKING-STORAGE SECTION.
       01 Var1 pic X(80) value 'a
      -     'aaa'. 
       END PROGRAM MyPGM.
```
Change it to:
```cobol
       IDENTIFICATION DIVISION.
       PROGRAM-ID. MyPGM.
       DATA DIVISION.
       WORKING-STORAGE SECTION.
       01 Var1 pic X(80) value 'a
      -    'aaa
      -    'bbb'.        
       END PROGRAM MyPGM.
```

## Test 3
From this code
```cobol
       IDENTIFICATION DIVISION.
       PROGRAM-ID. MyPGM.
       DATA DIVISION.
       WORKING-STORAGE SECTION.
       01 Var1 pic X(80) value 'a
      -     'aaa'. 
       END PROGRAM MyPGM.
```
Delete characters to get:
```cobol
       IDENTIFICATION DIVISION.
       PROGRAM-ID. MyPGM.
       DATA DIVISION.
       WORKING-STORAGE SECTION.
       01 Var1 pic X(80) value 'a'.
       END PROGRAM MyPGM.
```
No error expected.

## Test 4
From this code:
```cobol
       IDENTIFICATION DIVISION.
       PROGRAM-ID. MyPGM.
       DATA DIVISION.
       WORKING-STORAGE SECTION.
       01 Var1 pic X(80) value 'a
      -    'aaa
      -    'bbb'.        
       END PROGRAM MyPGM.
```
Change it to:
```cobol
       IDENTIFICATION DIVISION.
       PROGRAM-ID. MyPGM.
       DATA DIVISION.
       WORKING-STORAGE SECTION.
       01 Var1 pic X(80) value 'a
      -    'aaa'.      
       END PROGRAM MyPGM.
```

**Technical**
Will be edited by a team member.
Implementation details with the chosen solution.

**How to test automatically**
Use incremental tests.

### PR DESCRIPTION
**Pull request by fm-117**
Fixes #1673.

The solution used here will likely be updated when working on #2364 and its children.

### CODE REVIEW

**Review by smedilol — CHANGES_REQUESTED** [review]

File: TypeCobol/Compiler/Scanner/ScannerStep.cs [hunk 1]
```diff
@@ -270,28 +270,26 @@ private static void ScanTokensLineWithMultilineScanState(int lineToScanIndex, To
...
-                            else
+                            // Add into group
+                            lineToScan.SourceTokens.Clear();
+                            lineToScan.ScannerDiagnostics.Clear(); // TODO Move diagnostic clear operations into prepareDocumentLineForUpdate ?
+                            continuationLinesGroup.Insert(0, (TokensLine)prepareDocumentLineForUpdate(firstLineIndex, lineToScan, CompilationStep.Scanner));
+                            if (lineToScan.Type == CobolTextLineType.Source)
                             {
+                                // Found continuation group start -> break
                                 break;
                             }
```
[the remark this answers is not in the document]
Comment (fm-117) [reply TypeCobol/Compiler/Scanner/ScannerStep.cs:294]: This won't work as we do not have ScanState yet for all lines participating in the group.
We can use the `scanState` provided to the method but it won't work for special case involving REPLACE directives. The infamous combo : REPLACE + CONTINUATION + DEBUG + INCREMENTAL !

[the remark this answers is not in the document]
Comment (fm-117) [reply TypeCobol/Compiler/Scanner/ScannerStep.cs:294]: Also why not add the line no matter their type in the group ? I mean not scanning Invalid / Blank / Comment is ok but the scanner will handle them correctly anyway and we have to re-scan MultiFormalizedComment so isn't it better to add the lines without checking their type ?

[the remark this answers is not in the document]
Comment (fm-117) [reply TypeCobol/Compiler/Scanner/ScannerStep.cs:294]: Proposal:
[patch.txt](https://github.com/TypeCobolTeam/TypeCobol/files/10741342/patch.txt)

[the remark this answers is not in the document]
Comment (smedilol) [reply TypeCobol/Compiler/Scanner/ScannerStep.cs:294]: As discussed together, ok to keep the code as first proposed in your PR.
We will eventually handle a mix of Debug line and continuation line in #2446.

**Review by smedilol — APPROVED** [review]

**Pull request merged**